[ADD] resource_booking: new app to manage bookings

yajo opened the pull request:
This module adds a new app to allow you to book resource combinations in given
schedules.

Example use cases:

* Management of consultations in a clinic.
* Salesman appointments.
* Classroom and projector reservations.
* Hotel room booking.

Among the things you can do:

* Specify the type of booking, which includes a calendar of availability.
* Specify which resources can be booked together. All of them must be free to be booked.
* Place pending bookings, effectively giving permissions to someone to see the availability calendar and choose one slot.
* Partners can do that from their portals.
* If a partner has no user, he can still do the same via a tokenized URL.
* Backend users can also do that from the backend.
* Booking lifecycle with computed states.
* Automatic meeting creation and deletion.
* Automatic conflict detection.
* Deadline to block modifications.

@Tecnativa TT28201

sergio-teruel commented:
very red!! :smile:

yajo commented:
It should be less red now, thanks!

joao-p-marques commented on resource_booking/demo/res_users_demo.xml:
```diff
@@ -0,0 +1,11 @@
+
+<data>
+
+    <record id="base.user_demo" model="res.users">
+        <field name="groups_id" eval="[(4, ref('group_user'))]" />
+    </record>
```
Isn't this added by default?

yajo replied:
Nope, I'm making "Marc Demo" a resource booking user. Notice the ref is `group_user` and not `base.group_user`. When not indicating the module, it assumes the current one and translates into `resource_booking.group_user`.

joao-p-marques replied:
Oh ok, of course. In that case, maybe a more specific name to the group wouldn't harm, but that is already nitpicking, not important :+1:

joao-p-marques commented on resource_booking/models/calendar_event.py:
```diff
@@ -0,0 +1,53 @@
+        for (old_start, old_stop), new in zip(before, self):
+            if old_start == new.start and old_stop == new.stop:
+                rescheduled -= new
+        rescheduled._validate_booking_modifications()
```
If this throws an exception, isn't it being thrown after the write in line 47?
If you are not allowed to reschedule it, shouldn't the error abort the write?

yajo replied:
Yes, the transaction will be rolled back, so the final behavior is the one you say.

joao-p-marques commented on resource_booking/models/resource_booking_combination.py:
```diff
@@ -0,0 +1,101 @@
+        """Get available intervals for this booking combination."""
+        base = Intervals([(start_dt, end_dt, self)])
+        result = Intervals([])
+        for combination in self:
+            combination_intervals = base
+            for res in combination.resource_ids:
+                if not combination_intervals:
+                    break  # Can't restrict more
```
Why putting this check inside the loop? You could avoid entering that `if` in every iteration by:
```suggestion
        for combination in self:
            combination_intervals = base
            if combination_intervals:
                for res in combination.resource_ids:
```

yajo replied:
Yes, but `combination_intervals` gets more restricted with each loop (see next lines). If, after some iteration, it becomes void, we don't need to continue checking.

joao-p-marques replied:
Ok, thanks, I hadn't seen that `combination_intervals` changed bellow...

joao-p-marques commented on resource_booking/views/resource_booking_views.xml:
```diff
@@ -0,0 +1,130 @@
+                <field name="name" />
+                <field name="partner_id" avatar_field="image_small" />
+                <field name="combination_id" />
+                <!-- <field name="is_highlighted" invisible="1" /> -->
```
Should this be removed?

joao-p-marques reviewed (commented):
Code and functional review :+1: 

Just some tiny comments. :smiley: But overall LGTM

Commit:
[ADD] resource_booking: new app to manage bookings
This module adds a new app to allow you to book resource combinations in given
schedules.

Example use cases:

* Management of consultations in a clinic.
* Salesman appointments.
* Classroom and projector reservations.
* Hotel room booking.

Among the things you can do:

* Specify the type of booking, which includes a calendar of availability.
* Specify which resources can be booked together. All of them must be free to be booked.
* Place pending bookings, effectively giving permissions to someone to see the availability calendar and choose one slot.
* Partners can do that from their portals.
* If a partner has no user, he can still do the same via a tokenized URL.
* Backend users can also do that from the backend.
* Booking lifecycle with computed states.
* Automatic meeting creation and deletion.
* Automatic conflict detection.
* Deadline to block modifications.

@Tecnativa TT28201

yajo commented:
All attended/responded.

joao-p-marques reviewed (approved):
> All attended/responded.

Ok, thanks for the pointers :+1: 

Looks good now

yajo commented:
/ocabot merge nobump
